server/src/main/java/io/druid/server/BrokerQueryResource.java

```diff
@@ -0,0 +1,166 @@
+  ) throws IOException
+  {
+    final String reqContentType = req.getContentType();
+    final boolean isSmile = SmileMediaTypes.APPLICATION_JACKSON_SMILE.equals(reqContentType)
+                            || APPLICATION_SMILE.equals(reqContentType);
+    final String contentType = isSmile ? SmileMediaTypes.APPLICATION_JACKSON_SMILE : MediaType.APPLICATION_JSON;
+
+    ObjectMapper objectMapper = isSmile ? smileMapper : jsonMapper;
+    final ObjectWriter jsonWriter = pretty != null
+                                    ? objectMapper.writerWithDefaultPrettyPrinter()
+                                    : objectMapper.writer();
+
+    try {
+      Query<?> query = objectMapper.readValue(in, Query.class);
+      List<LocatedSegmentDescriptor> located = getTargetLocations(query.getDataSource(), query.getIntervals());
+      if (located == null || located.isEmpty()) {
+        return Response.ok(Sequences.empty()).build();
+      }
+      return Response.ok(jsonWriter.writeValueAsString(located), contentType).build();
+    }
+    catch (Exception e) {
+      return Response.serverError().type(contentType).entity(
+          jsonWriter.writeValueAsString(
+              ImmutableMap.of(
+                  "error", e.getMessage() == null ? "null exception" : e.getMessage()
+              )
+          )
+      ).build();
+    }
+  }
+
+  @GET
+  @Path("/candidates")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getQueryTargets(
+      @QueryParam("datasource") String datasource,
+      @QueryParam("intervals") String intervals
+  ) throws IOException
+  {
+    List<Interval> intervalList = Lists.newArrayList();
+    for (String interval : intervals.split(",")) {
+      intervalList.add(Interval.parse(interval.trim()));
+    }
+    List<LocatedSegmentDescriptor> located = getTargetLocations(
+        new TableDataSource(datasource),
+        JodaUtils.condenseIntervals(intervalList)
+    );
+    return Response.ok(located).build();
+  }
+
+  private List<LocatedSegmentDescriptor> getTargetLocations(DataSource datasource, List<Interval> intervals)
+  {
+    TimelineLookup<String, ServerSelector> timeline = brokerServerView.getTimeline(datasource);
+    if (timeline == null) {
+      return Collections.emptyList();
+    }
+    List<LocatedSegmentDescriptor> located = Lists.newArrayList();
+    for (Interval interval : intervals) {
+      for (TimelineObjectHolder<String, ServerSelector> holder : timeline.lookup(interval)) {
+        for (PartitionChunk<ServerSelector> chunk : holder.getObject()) {
+          ServerSelector selector = chunk.getObject();
+          final SegmentDescriptor descriptor = new SegmentDescriptor(
+              holder.getInterval(), holder.getVersion(), chunk.getChunkNumber()
+          );
+          List<DruidServerMetadata> candidates = selector.getCandidates();
+          located.add(new LocatedSegmentDescriptor(descriptor, holder.getApproximatedSize(), candidates));
```

Comment:
I am not sure if `getApproximatedSize()` is needed. You can get the size of a segment by calling `selector.getSegment.getSize()`. Adding size to TimelineObjectHolder could increase Broker memory usage.